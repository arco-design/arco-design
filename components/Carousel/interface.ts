import { CSSProperties, MutableRefObject, ReactNode } from 'react';

export type CarouselHandle = {
  dom: HTMLElement;
  goto: (options: {
    /** 目标索引 */
    index: number;
    /** 是否为逆向 */
    isNegative?: boolean;
    /** 是否由用户触发，将决定 onChange 回调的第三个参数 */
    isManual?: boolean;
    /** 是否重置自动播放的 interval */
    resetAutoPlayInterval?: boolean;
  }) => void;
};

/**
 * @title Carousel
 */
export interface CarouselProps {
  style?: CSSProperties;
  className?: string | string[];
  children?: ReactNode;
  /**
   * @zh 当前展示索引
   * @en The index of current slide which starts from 0
   * @defaultValue 0
   */
  currentIndex?: number;
  /**
   * @zh
   * 是否自动循环展示，或者传入 `{ interval: 自动切换的时间间隔(默认: 3000), hoverToPause: 鼠标悬浮时是否暂停自动切换(默认: true) }` 进行高级配置 (`2.14.0` 支持传入对象)
   * @en
   * Whether to scroll automatically, or pass in `{ interval: the time interval for switching (default: 3000),
   * hoverToPause: whether to pause switching while hover (default: true) }` for configuration (object is supported from `2.14.0`)
   */
  autoPlay?: boolean | { interval?: number; hoverToPause?: boolean };
  /**
   * @zh
   * 是否仅渲染满足动画效果的最少数量的 children
   * @en
   * Whether to render only the minimum number of children that meet the animation effect
   * @version 2.21.0
   */
  miniRender?: boolean;
  // TODO: 3.x 移除此参数
  autoPlaySpeed?: number;
  /**
   * @zh 幻灯片移动速率(ms)
   * @en The duration of the slide movement(ms)
   * @defaultValue 500
   */
  moveSpeed?: number;
  /**
   * @zh 切换动画
   * @en The animation of the slide movement
   * @defaultValue slide
   */
  animation?: 'slide' | 'card' | 'fade';
  /**
   * @zh 幻灯片切换触发方式, click/hover 指示器
   * @en How to trigger the slide switch, click/hover the indicator
   * @defaultValue click
   */
  trigger?: 'click' | 'hover';
  /**
   * @zh 幻灯片移动方向
   * @en The direction of the slide movement
   * @defaultValue horizontal
   */
  direction?: 'horizontal' | 'vertical';
  /**
   * @zh 切换箭头显示时机
   * @en When to show the switch trigger
   * @defaultValue always
   */
  showArrow?: 'always' | 'hover' | 'never';
  /**
   * @zh 切换箭头样式
   * @en The additional css class for switch trigger
   */
  arrowClassName?: string | string[];
  /**
   * @zh 自定义图标
   * @en Customize icons
   * @version 2.25.0
   */
  icons?: {
    prev?: ReactNode;
    next?: ReactNode;
  };
  /**
   * @zh 指示器类型，可为小方块和小圆点或不显示
   * @en Type of indicator
   * @defaultValue dot
   */
  indicatorType?: 'line' | 'dot' | 'slider' | 'never';
  /**
   * @zh 指示器位置
   * @en Position of indication
   * @defaultValue bottom
   */
  indicatorPosition?: 'bottom' | 'top' | 'left' | 'right' | 'outer';
  /**
   * @zh 指示器的样式
   * @en The additional css class for indicator
   */
  indicatorClassName?: string | string[];
  /**
   * @zh
   * 过渡速度曲线, 默认匀速 [transition-timing-function](https://developer.mozilla.org/zh-CN/docs/Web/CSS/transition-timing-function)
   * @en
   * How intermediate values are calculated for CSS properties being affected by a transition effect.
   * [transition-timing-function](https://developer.mozilla.org/zh-CN/docs/Web/CSS/transition-timing-function)
   * @defaultValue cubic-bezier(0.34, 0.69, 0.1, 1)
   */
  timingFunc?: string;
  /**
   * @zh 幻灯片发生切换时的回调函数。
   * @en Callback when slide changes.
   * @version `isManual` in 2.4.0
   */
  onChange?: (index: number, prevIndex: number, isManual: boolean) => void;
  /**
   * @zh 用于获得带有 API 方法的 Carousel 引用。
   * @en Carousel reference for imperative API calls.
   * @version 2.16.1
   */
  carousel?: MutableRefObject<CarouselHandle>;
}

export interface CarouselArrowProps
  extends Pick<CarouselProps, 'direction' | 'showArrow' | 'icons'> {
  className?: string | string[];
  prev?: (e: any) => any;
  next?: (e: any) => any;
}

export interface CarouselIndicatorProps {
  className?: string | string[];
  count: number;
  activeIndex?: number;
  type?: CarouselProps['indicatorType'];
  position?: CarouselProps['indicatorPosition'];
  trigger?: CarouselProps['trigger'];
  onSelectIndex?: (e: number) => void;
}
