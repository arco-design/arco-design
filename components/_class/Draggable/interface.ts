import { CSSProperties, ReactNode } from 'react';

export type DragStatus = 'none' | 'dragged' | 'dragging';

export type DragPosition = 'left' | 'right' | 'top' | 'bottom';

export interface DraggableItemProps extends Pick<DraggableProps, 'direction'> {
  style?: CSSProperties;
  prefixCls: string;
  children?: ReactNode;
  /** Weather allow to drag  */
  disabled?: boolean;
  /** Weather allow to drop on it */
  droppable?: boolean;
  onDragStart?: (event) => void;
  onDragEnd?: (event) => void;
  onDragLeave?: (event) => void;
  onDragOver?: (event) => void;
  onDrop?: (event, dropPosition: DragPosition) => void;
}

export interface DraggableProps {
  children?: ReactNode;
  direction?: 'horizontal' | 'vertical';
  className?: string | string[];
  itemWrapperStyle?: CSSProperties;
  onIndexChange?: (index: number, prevIndex: number) => void;
}
