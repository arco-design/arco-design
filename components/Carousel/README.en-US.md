`````
Component / Data Display

# Carousel

Carousel is used to display multiple contents such as pictures, videos, or embedded frames on a rotating timer. It supports both automatic playback and manual switching.
`````

%%Content%%

## API

### Carousel

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|style|Additional style|`CSSProperties`|`-`|-|
|className|Additional css class|`string \| string[]`|`-`|-|
|currentIndex|The index of current slide which starts from 0|`number`|`0`|-|
|autoPlay|Whether to scroll automatically, or pass in `{ interval: the time interval for switching (default: 3000),hoverToPause: whether to pause switching while hover (default: true) }` for configuration (object is supported from `2.14.0`)|`boolean \| { interval?: number; hoverToPause?: boolean }`|`-`|-|
|miniRender|Whether to render only the minimum number of children that meet the animation effect|`boolean`|`-`|2.21.0|
|moveSpeed|The duration of the slide movement(ms)|`number`|`500`|-|
|animation|The animation of the slide movement|`'slide' \| 'card' \| 'fade'`|`slide`|-|
|trigger|How to trigger the slide switch, click/hover the indicator|`'click' \| 'hover'`|`click`|-|
|direction|The direction of the slide movement|`'horizontal' \| 'vertical'`|`horizontal`|-|
|showArrow|When to show the switch trigger|`'always' \| 'hover' \| 'never'`|`always`|-|
|arrowClassName|The additional css class for switch trigger|`string \| string[]`|`-`|-|
|icons|Customize icons|`{prev?: ReactNode;next?: ReactNode;}`|`-`|2.25.0|
|indicatorType|Type of indicator|`'line' \| 'dot' \| 'slider' \| 'never'`|`dot`|-|
|indicatorPosition|Position of indication|`'bottom' \| 'top' \| 'left' \| 'right' \| 'outer'`|`bottom`|-|
|indicatorClassName|The additional css class for indicator|`string \| string[]`|`-`|-|
|timingFunc|How intermediate values are calculated for CSS properties being affected by a transition effect.[transition-timing-function](https://developer.mozilla.org/zh-CN/docs/Web/CSS/transition-timing-function)|`string`|`cubic-bezier(0.34, 0.69, 0.1, 1)`|-|
|onChange|Callback when slide changes.|`(index: number, prevIndex: number, isManual: boolean) => void`|`-`|`isManual` in 2.4.0|
|carousel|Carousel reference for imperative API calls.|`MutableRefObject<CarouselHandle>`|`-`|2.16.1|
