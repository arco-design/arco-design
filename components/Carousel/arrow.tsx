import React, { useContext } from 'react';
import IconLeft from '../../icon/react-icon/IconLeft';
import IconRight from '../../icon/react-icon/IconRight';
import IconUp from '../../icon/react-icon/IconUp';
import IconDown from '../../icon/react-icon/IconDown';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { CarouselArrowProps } from './interface';
import useKeyboardEvent from '../_util/hooks/useKeyboardEvent';

function CarouselArrow(props: CarouselArrowProps, ref) {
  const { className, direction = 'horizontal', showArrow = 'always', prev, next, icons } = props;

  const getKeyboardEvents = useKeyboardEvent();
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('carousel');
  const arrowClass = cs(
    `${prefixCls}-arrow`,
    {
      [`${prefixCls}-arrow-hover`]: showArrow === 'hover',
    },
    className
  );
  const iconPrev =
    icons && icons.hasOwnProperty('prev') ? (
      icons.prev
    ) : direction === 'horizontal' ? (
      <IconLeft />
    ) : (
      <IconUp />
    );
  const iconNext =
    icons && icons.hasOwnProperty('next') ? (
      icons.next
    ) : direction === 'horizontal' ? (
      <IconRight />
    ) : (
      <IconDown />
    );

  return (
    <div ref={ref} className={arrowClass}>
      <div
        className={`${prefixCls}-arrow-${direction === 'vertical' ? 'top' : 'left'}`}
        onClick={prev}
        role="button"
        tabIndex={0}
        {...getKeyboardEvents({ onPressEnter: prev })}
      >
        {iconPrev}
      </div>
      <div
        className={`${prefixCls}-arrow-${direction === 'vertical' ? 'bottom' : 'right'}`}
        onClick={next}
        role="button"
        tabIndex={0}
        {...getKeyboardEvents({ onPressEnter: next })}
      >
        {iconNext}
      </div>
    </div>
  );
}

const CarouselArrowComponent = React.forwardRef<unknown, CarouselArrowProps>(CarouselArrow);

export default CarouselArrowComponent;
