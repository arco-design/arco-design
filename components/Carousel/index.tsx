import React, {
  CSSProperties,
  ReactElement,
  useContext,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import cs from '../_util/classNames';
import CarouselIndicator from './indicator';
import CarouselArrow from './arrow';
import { ConfigContext } from '../ConfigProvider';
import { CarouselProps } from './interface';
import { useInterval } from '../_util/hooks/useInterval';
import ResizeObserver from '../_util/resizeObserver';
import warning from '../_util/warning';
import omit from '../_util/omit';
import useMergeProps from '../_util/hooks/useMergeProps';

const DEFAULT_AUTO_PLAY_INTERVAL = 3000;

const defaultProps: CarouselProps = {
  animation: 'slide',
  indicatorType: 'dot',
  indicatorPosition: 'bottom',
  direction: 'horizontal',
  showArrow: 'always',
  trigger: 'click',
  moveSpeed: 500,
  timingFunc: 'cubic-bezier(0.34, 0.69, 0.1, 1)',
};

function Carousel(baseProps: CarouselProps, ref) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<CarouselProps>(baseProps, defaultProps, componentConfig?.Carousel);
  const {
    style,
    className,
    children,
    currentIndex,
    animation,
    trigger,
    direction,
    moveSpeed,
    timingFunc,
    indicatorType,
    indicatorPosition,
    indicatorClassName,
    showArrow,
    miniRender,
    arrowClassName,
    carousel,
    icons,
    onChange,
    ...rest
  } = props;

  // TODO 兼容 autoPlaySpeed， 3.x 移除
  let autoPlay = props.autoPlay;
  if (autoPlay && props.autoPlaySpeed) {
    autoPlay = {
      interval: props.autoPlaySpeed,
    };
    warning(
      true,
      `[Arco Carousel] The 'autoPlaySpeed' property will be removed in the next major version, please use 'autoPlay.interval' instead.`
    );
  }

  const childrenList = React.Children.toArray(children).filter((child) =>
    React.isValidElement(child)
  ) as ReactElement[];
  const childrenLength = childrenList.length;

  const refDom = useRef(null);
  const refSliderWrapper = useRef(null);
  const refAnimationTimer = useRef(null);

  const [index, setIndex] = useState(
    typeof currentIndex === 'number' ? getValidIndex(currentIndex) : 0
  );
  const [previousIndex, setPreviousIndex] = useState<number>(index);
  const [isPause, setIsPause] = useState(false);
  const [isAnimating, setIsAnimating] = useState(false);
  const [slideDirection, setSlideDirection] = useState<'positive' | 'negative'>(null);
  const [computedStyle, setComputedStyle] = useState<{
    sliderWrapper: CSSProperties;
    indicatorWrapper: CSSProperties;
  }>({
    sliderWrapper: null,
    indicatorWrapper: null,
  });

  const mergedIndex = typeof currentIndex === 'number' ? getValidIndex(currentIndex) : index;
  const prevIndex = getValidIndex(mergedIndex - 1);
  const nextIndex = getValidIndex(mergedIndex + 1);
  const autoPlayInterval =
    typeof autoPlay === 'object' && autoPlay.interval
      ? autoPlay.interval
      : DEFAULT_AUTO_PLAY_INTERVAL;

  useEffect(() => {
    return () => refAnimationTimer.current && clearTimeout(refAnimationTimer.current);
  }, []);

  useEffect(() => {
    slideTo({
      targetIndex: mergedIndex,
    });
  }, [mergedIndex]);

  const resetInterval = useInterval(
    () => {
      slideTo({
        targetIndex: nextIndex,
      });
    },
    autoPlay && !isPause && childrenLength > 1 ? autoPlayInterval : null
  );

  useImperativeHandle(carousel, () => {
    return {
      dom: refDom.current,
      goto: ({ index, isNegative, isManual, resetAutoPlayInterval }) => {
        slideTo({
          targetIndex: getValidIndex(index),
          isNegative,
          isManual,
          resetAutoPlayInterval,
        });
      },
    };
  });

  function getValidIndex(i): number {
    const indexNumber = +i;
    return typeof indexNumber === 'number' && !isNaN(indexNumber)
      ? (indexNumber + childrenLength) % childrenLength
      : i;
  }

  function slideTo({
    targetIndex,
    isNegative = false,
    isManual = false,
    resetAutoPlayInterval = false,
  }: {
    targetIndex: number;
    isNegative?: boolean;
    isManual?: boolean;
    resetAutoPlayInterval?: boolean;
  }) {
    if (!isAnimating && targetIndex !== mergedIndex) {
      setIsAnimating(true);
      setIndex(targetIndex);
      setPreviousIndex(index);
      setSlideDirection(isNegative ? 'negative' : 'positive');
      onChange && onChange(targetIndex, mergedIndex, isManual);

      if (autoPlay && resetAutoPlayInterval) {
        resetInterval();
      }

      refAnimationTimer.current = setTimeout(() => {
        setIsAnimating(false);
        refAnimationTimer.current = null;
      }, moveSpeed);
    }
  }

  function computeStyle() {
    if (animation === 'card') {
      if (refSliderWrapper.current) {
        const sliderElement = refSliderWrapper.current.children[0];

        if (!sliderElement) {
          return;
        }

        const totalWidth = refSliderWrapper.current.clientWidth;
        const sliderWidth = sliderElement.clientWidth;
        const edge = (totalWidth - sliderWidth) / 2;

        // deltaZ is TranslateZ(-Zpx) of prev/next slider's style
        // perspective / (perspective + deltaZ) = x / X
        const deltaZ = 200;
        const x = totalWidth / 2;
        const X = sliderWidth;
        // avoid getting a huge perspective value
        const perspective = x + 50 >= X ? deltaZ * 4 : (deltaZ * x) / (X - x);

        setComputedStyle({
          sliderWrapper: {
            perspective,
          },
          indicatorWrapper: {
            width: 'auto',
            left: edge,
            right: edge,
          },
        });
      }
    } else {
      setComputedStyle({
        sliderWrapper: null,
        indicatorWrapper: null,
      });
    }
  }

  const prefixCls = getPrefixCls('carousel');
  const classNames = cs(
    prefixCls,
    `${prefixCls}-indicator-position-${indicatorPosition}`,
    className
  );
  const eventHandlers = Object.assign(
    {},
    autoPlay && (typeof autoPlay !== 'object' || autoPlay.hoverToPause !== false)
      ? {
          onMouseEnter: () => setIsPause(true),
          onMouseLeave: () => setIsPause(false),
        }
      : null
  );

  return (
    <ResizeObserver onResize={computeStyle}>
      <div
        ref={(_ref) => {
          ref = _ref;
          refDom.current = ref;
        }}
        className={classNames}
        style={style}
        {...omit(rest, ['autoplay', 'autoPlaySpeed'])}
        {...eventHandlers}
      >
        <div
          ref={refSliderWrapper}
          style={computedStyle.sliderWrapper}
          className={cs(`${prefixCls}-${animation}`, `${prefixCls}-${direction}`, {
            [`${prefixCls}-negative`]: slideDirection === 'negative',
          })}
        >
          {childrenList.map((child, index) => {
            const isCurrent = index === mergedIndex;
            const isPrev = index === prevIndex;
            const isNext = index === nextIndex;
            const shouldRenderChild = !miniRender || isCurrent || isPrev || isNext;

            if (!shouldRenderChild) {
              return null;
            }

            const {
              style: childStyle,
              className: childClassName,
              onClick: childOnClick,
            } = child.props;

            return React.cloneElement(child, {
              style: Object.assign(
                {
                  transitionTimingFunction: timingFunc,
                  transitionDuration: `${moveSpeed}ms`,
                  animationTimingFunction: timingFunc,
                  animationDuration: `${moveSpeed}ms`,
                },
                childStyle
              ),
              className: cs(childClassName, {
                [`${prefixCls}-item-prev`]: isPrev,
                [`${prefixCls}-item-next`]: isNext,
                [`${prefixCls}-item-current`]: isCurrent,
                [`${prefixCls}-item-slide-in`]:
                  animation === 'slide' && slideDirection && isAnimating && isCurrent,
                [`${prefixCls}-item-slide-out`]:
                  animation === 'slide' && slideDirection && isAnimating && index === previousIndex,
              }),
              onClick: (event) => {
                childOnClick && childOnClick(event);
                slideTo({
                  targetIndex: index,
                  isNegative: index === prevIndex,
                  isManual: true,
                });
              },
            });
          })}
        </div>

        {indicatorType !== 'never' && childrenLength > 1 && (
          <div
            style={computedStyle.indicatorWrapper}
            className={cs(
              `${prefixCls}-indicator-wrapper`,
              `${prefixCls}-indicator-wrapper-${indicatorPosition}`
            )}
          >
            <CarouselIndicator
              className={indicatorClassName}
              type={indicatorType}
              count={childrenLength}
              activeIndex={mergedIndex}
              position={indicatorPosition}
              trigger={trigger}
              onSelectIndex={(index) =>
                slideTo({
                  targetIndex: index,
                  isNegative: index < mergedIndex,
                  isManual: true,
                })
              }
            />
          </div>
        )}

        {showArrow !== 'never' && childrenLength > 1 && (
          <CarouselArrow
            className={arrowClassName}
            direction={direction}
            showArrow={showArrow}
            icons={icons}
            prev={() =>
              slideTo({
                targetIndex: prevIndex,
                isNegative: true,
                isManual: true,
              })
            }
            next={() =>
              slideTo({
                targetIndex: nextIndex,
                isManual: true,
              })
            }
          />
        )}
      </div>
    </ResizeObserver>
  );
}

const CarouselComponent = React.forwardRef<unknown, CarouselProps>(Carousel);

CarouselComponent.displayName = 'Carousel';

export default CarouselComponent;

export { CarouselProps };
