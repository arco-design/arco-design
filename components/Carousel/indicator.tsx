import React, { ReactNode, useContext } from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { CarouselIndicatorProps } from './interface';

function CarouselIndicator(props: CarouselIndicatorProps, ref) {
  const { getPrefixCls } = useContext(ConfigContext);
  const {
    className,
    type = 'line',
    count = 2,
    activeIndex = 0,
    position = 'bottom',
    trigger = 'click',
    onSelectIndex,
  } = props;

  const prefixCls = getPrefixCls('carousel-indicator');
  const indicatorContent: ReactNode[] = [];

  if (type === 'slider') {
    const step = 100 / count;
    indicatorContent.push(
      <span
        key={0}
        style={{ width: `${step}%`, left: `${activeIndex * step}%` }}
        className={cs(`${prefixCls}-item`, `${prefixCls}-item-active`)}
      />
    );
  } else {
    for (let i = 0; i < count; i++) {
      indicatorContent.push(
        <span
          key={i}
          data-index={i}
          className={cs(`${prefixCls}-item`, {
            [`${prefixCls}-item-active`]: i === activeIndex,
          })}
        />
      );
    }
  }

  const wrapperProps = {
    ref,
    className: cs(prefixCls, `${prefixCls}-${type}`, `${prefixCls}-${position}`, className),
    [trigger === 'click' ? 'onClick' : 'onMouseMove']: (event) => {
      event.preventDefault();
      if (type === 'slider') {
        // clear up effect from event bubbling
        if (event.target === event.currentTarget) {
          const { x: startX, width } = event.currentTarget.getBoundingClientRect();
          const offsetX = event.nativeEvent.clientX - startX;
          const index = ~~((offsetX / width) * count);
          index !== activeIndex && onSelectIndex(index);
        }
      } else {
        const dataIndex: string = event.target.getAttribute('data-index');
        // Judge if data-index exists at first, event.target can be the wrapper of indicators
        dataIndex && +dataIndex !== activeIndex && onSelectIndex(+dataIndex);
      }
    },
  };

  return <div {...wrapperProps}>{indicatorContent}</div>;
}

const CarouselIndicatorComponent = React.forwardRef<unknown, CarouselIndicatorProps>(
  CarouselIndicator
);

export default CarouselIndicatorComponent;
