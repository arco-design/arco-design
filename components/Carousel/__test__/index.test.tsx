import React from 'react';
import { mount, render } from 'enzyme';
import { act } from 'react-dom/test-utils';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Carousel from '..';
import { CarouselHandle } from '../interface';
import { sleep } from '../../../tests/util';

jest.mock('resize-observer-polyfill', () => ({
  __esModule: true,
  default: jest.fn().mockImplementation((cb) => {
    // 直接触发，不用onResize
    cb();

    return {
      observe: jest.fn(),
      unobserve: jest.fn(),
      disconnect: jest.fn(),
    };
  }),
}));

mountTest(Carousel);
componentConfigTest(Carousel, 'Carousel');

const prefixCls = '.arco-carousel';

const wrapperStyle: React.CSSProperties = {
  width: '600px',
  height: '300px',
};

const itemStyle1: React.CSSProperties = {
  backgroundColor: '#3370FF',
  color: '#ffffff',
  textAlign: 'center',
  width: '100%',
  height: '100%',
  fontSize: '25px',
  padding: '100px',
};

const itemStyle2: React.CSSProperties = {
  backgroundColor: '#ADCEFF',
  color: '#ffffff',
  textAlign: 'center',
  width: '100%',
  height: '100%',
  fontSize: '25px',
  padding: '100px',
};

const isActiveIndex = (wrapper, index) => {
  jest.advanceTimersByTime(50);
  wrapper.update();
  const items = wrapper.find('.arco-carousel-indicator-item');
  return items.at(index).hasClass('arco-carousel-indicator-item-active');
};

describe('Carousel', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render only child correctly', () => {
    const wrapper = render(
      <Carousel style={wrapperStyle}>
        <div style={itemStyle1} />
      </Carousel>
    );
    const slider = wrapper.find(`${prefixCls}-slider`);
    expect(slider.children).toHaveLength(1);
  });

  it('should init currentIndex correctly', () => {
    const currentIndex = 3;
    const wrapper = mount(
      <Carousel style={wrapperStyle} currentIndex={currentIndex}>
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
        <div style={itemStyle2} />
      </Carousel>
    );
    expect(isActiveIndex(wrapper, 3)).toBe(true);
  });

  it('should autoPlay correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} autoPlay>
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
        <div style={itemStyle2} />
      </Carousel>
    );
    jest.advanceTimersByTime(3000);
    expect(isActiveIndex(wrapper, 1)).toBe(true);
  });

  it('miniRender works correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} miniRender>
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
      </Carousel>
    );
    expect(wrapper.find('.arco-carousel-slide > div')).toHaveLength(3);
  });

  it('should move correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle}>
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
      </Carousel>
    );

    wrapper.find('.arco-carousel-arrow-right').at(0).simulate('click');
    expect(isActiveIndex(wrapper, 1)).toBe(true);

    // 等待向右的动画结束，否则下次点击会被忽略
    jest.advanceTimersByTime(1000);

    wrapper.find('.arco-carousel-arrow-left').at(0).simulate('click');
    expect(isActiveIndex(wrapper, 0)).toBe(true);
  });

  it('should selectIndex correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle}>
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
      </Carousel>
    );
    wrapper.find('.arco-carousel-indicator-item').at(2).simulate('click');
    expect(isActiveIndex(wrapper, 2)).toBe(true);
  });

  it('should addInterval, clearInterval correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} autoPlay>
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
        <div style={itemStyle2} />
      </Carousel>
    );

    wrapper.simulate('mouseenter');
    jest.advanceTimersByTime(4000);
    expect(isActiveIndex(wrapper, 0)).toBe(true);
    wrapper.simulate('mouseleave');
    jest.advanceTimersByTime(4000);
    expect(isActiveIndex(wrapper, 1)).toBe(true);
  });

  it('should be vertical', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} direction="vertical" indicatorPosition="right">
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
      </Carousel>
    );
    expect(wrapper.find(`${prefixCls}-vertical`)).toHaveLength(1);
  });

  it('should be dots', () => {
    const wrapper = mount(
      <Carousel
        style={wrapperStyle}
        direction="vertical"
        indicatorType="dot"
        indicatorPosition="right"
        showArrow="hover"
      >
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
      </Carousel>
    );
    expect(wrapper.find('.arco-carousel-indicator-dot')).toHaveLength(1);
    expect(wrapper.find('.arco-carousel-indicator-right')).toHaveLength(1);
    expect(wrapper.find('.arco-carousel-arrow').hasClass('arco-carousel-arrow-hover')).toBe(true);
  });

  it('should self timingfunc correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} moveSpeed={800} timingFunc="ease-out">
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
      </Carousel>
    );

    const slider = wrapper.find(`${prefixCls}-slide`);
    const style = slider.childAt(0).prop('style');
    expect(style.transitionTimingFunction).toEqual('ease-out');
    expect(style.transitionDuration).toEqual('800ms');
  });

  it('support the ref trigger change scene', async () => {
    jest.useRealTimers();
    const carousel = React.createRef<CarouselHandle>();
    const onChange = jest.fn();
    const wrapper = mount(
      <Carousel
        style={wrapperStyle}
        autoPlay={{ interval: 1000 }}
        carousel={carousel}
        onChange={onChange}
      >
        <div style={itemStyle1} />
        <div style={itemStyle2} />
        <div style={itemStyle1} />
        <div style={itemStyle2} />
      </Carousel>
    );

    const { dom, goto } = carousel.current;
    expect(dom.className.includes('arco-carousel ')).toBeTruthy();
    expect(typeof goto).toBe('function');
    await act(async () => {
      goto({ index: 2, resetAutoPlayInterval: true });
      wrapper.update();
    });

    expect(isActiveIndex(wrapper, 2)).toBe(true);
    expect(onChange.mock.calls).toHaveLength(1);
    expect(onChange.mock.calls[0]).toEqual([2, 0, false]);

    await act(async () => {
      await sleep(1000);
      wrapper.update();
    });

    expect(onChange.mock.calls[1]).toEqual([3, 2, false]);

    await act(async () => {
      goto({ index: 1, isManual: true, isNegative: true });
      wrapper.update();
    });
    expect(onChange).toHaveBeenCalledTimes(3);
    expect(isActiveIndex(wrapper, 1)).toBe(true);
  });

  it('support child props scene', () => {
    const handleClick = jest.fn();
    const wrapper = mount(
      <Carousel style={wrapperStyle}>
        <li style={itemStyle1} />
        <li style={itemStyle2} className="carousel-child" />
        <li style={itemStyle1} onClick={handleClick} />
        <li style={itemStyle2} />
      </Carousel>
    );

    act(() => {
      wrapper.find('li').at(2).simulate('click');

      wrapper.update();
    });
    expect(handleClick).toHaveBeenCalled();
    expect(isActiveIndex(wrapper, 2)).toBeTruthy();

    expect(wrapper.find('li').at(1).hasClass('carousel-child')).toBeTruthy();
  });

  it('support card animation scene', async () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} animation="card">
        <div style={{ width: '60%' }}>
          <div style={itemStyle1} />
        </div>
        <div style={{ width: '60%' }}>
          <div style={itemStyle2} />
        </div>
        <div style={{ width: '60%' }}>
          <div style={itemStyle1} />
        </div>
      </Carousel>
    );
    const cardWrapper = wrapper.find(`${prefixCls}-card`);
    expect(cardWrapper).toHaveLength(1);

    expect(cardWrapper.prop('style').perspective).not.toBeUndefined();
  });

  it('support slider indicatorType correctly', () => {
    const wrapper = mount(
      <Carousel style={wrapperStyle} autoPlay indicatorType="slider">
        {new Array(4).fill('1').map((_, i) => (
          <div key={i} style={itemStyle1} />
        ))}
      </Carousel>
    );

    expect(wrapper.find('.arco-carousel-indicator-slider')).toHaveLength(1);
    act(() => {
      jest.advanceTimersByTime(3000);
      wrapper.update();
    });
    const activeIndicator = wrapper
      .find('.arco-carousel-indicator-slider')
      .find('.arco-carousel-indicator-item');
    expect(activeIndicator.prop('style')).toEqual({
      width: `${100 / 4}%`,
      left: `${100 / 4}%`,
    });
  });

  it('customize icon in arrows', () => {
    const wrapper = render(
      <Carousel style={wrapperStyle} icons={{ prev: null, next: '>' }}>
        <li style={itemStyle1} />
        <li style={itemStyle2} />
      </Carousel>
    );
    expect(wrapper.find(`${prefixCls}-arrow-left`).text()).toBe('');
    expect(wrapper.find(`${prefixCls}-arrow-right`).text()).toBe('>');
  });
});
