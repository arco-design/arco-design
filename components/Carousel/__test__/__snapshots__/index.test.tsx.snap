// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Carousel default 1`] = `
<div
  class="arco-carousel arco-carousel-indicator-position-bottom"
>
  <div
    class="arco-carousel-slide arco-carousel-horizontal"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Carousel set className globally 1`] = `
<div
  class="arco-carousel arco-carousel-indicator-position-bottom global-default-class"
>
  <div
    class="arco-carousel-slide arco-carousel-horizontal"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Carousel set className globally and component set className itself. 1`] = `
<div
  class="arco-carousel arco-carousel-indicator-position-bottom component-class"
>
  <div
    class="arco-carousel-slide arco-carousel-horizontal"
  />
</div>
`;

exports[`ConfigProvider componentConfig.Carousel set data-* property 1`] = `
<div
  class="arco-carousel arco-carousel-indicator-position-bottom"
  data-test-name="Carousel"
>
  <div
    class="arco-carousel-slide arco-carousel-horizontal"
  />
</div>
`;
